src/units/abilities.cpp

```diff
@@ -867,41 +867,118 @@ static void add_name(std::string& weapon_abilities, bool active, const config::a
 std::string attack_type::weapon_specials(bool only_active, bool is_backstab) const
 {
 	//log_scope("weapon_specials");
-	std::string res;
-	for (const config::any_child sp : specials_.all_children_range())
-	{
-		const bool active = special_active(sp.cfg, AFFECT_EITHER, sp.key, is_backstab);
+	return weapon_specials_impl(only_active, is_backstab);
+}
 
-		const std::string& name =
-			active
-			? sp.cfg["name"].str()
-			: default_value(sp.cfg, "name_inactive", "name").str();
-		if (!name.empty()) {
-			if (!res.empty()) res += ", ";
-			if (only_active && !active) res += font::span_color(font::inactive_details_color);
-			res += name;
-			if (only_active && !active) res += "</span>";
+std::string attack_type::weapon_specials_value(const std::set<std::string> checking_tags, bool only_active, bool is_backstab) const
+{
+	//log_scope("weapon_specials_value");
+	return weapon_specials_impl(only_active, is_backstab, checking_tags, true);
+}
+
+std::string attack_type::weapon_specials_impl(bool only_active, bool is_backstab, const std::set<std::string> checking_tags, bool is_tooltip) const
+{
+	std::string res;
+	for (const config::any_child sp : specials_.all_children_range()) {
+		bool tag_checked = is_tooltip ? (checking_tags.count(sp.key) != 0) : true;
+		if(tag_checked){
+			const bool active = is_tooltip ? special_active(sp.cfg, AFFECT_SELF, sp.key, is_backstab) : special_active(sp.cfg, AFFECT_EITHER, sp.key, is_backstab);
+			const std::string& default_name = !is_tooltip ? default_value(sp.cfg, "name_inactive", "name").str() : "";
+			const std::string& name = active ? sp.cfg["name"].str() : default_name;
```

Comment:
Again, I'm not understanding why constructing a tooltip should be a reason to use a different value. That single boolean argument seems to be moving logic that should be in the caller into this function.

I think the solution should be to move the presentation into the caller - instead of returning a string here, make a `struct` just to be this function's return type. I'm thinking it would be something along the lines of:
```
	/** Part of the return structure for weapon_specials */
	struct name_and_active {
		/** Contains the special's name or name_inactive, depending on whether it's active */
		std::string name; // TODO: I assume this should be a tstring
		bool active;
	};
	struct RENAME_THIS_THING {
		std::vector<name_and_active> from_self;
		std::vector<name_and_active> from_own_teacher; // is this always 'teacher', or can it be other allies?
		std::vector<name_and_active> from_opponent;
		std::vector<name_and_active> from_enemy_teacher;
	};
```

With that, move the presentation logic into the caller. Everything about tracking whether the current section heading has already been used in the string moves there, and the code in abilities.cpp becomes simpler.

Reply:
except what using struct is not compatible with void add_name function, and the code don't seem becomes simpler.